File: 数据结构/Java/第5章 队列/5.5 双端队列/Deque.java
import java.util.List;
import java.util.LinkedList;

public class Deque<E> {
	private List<E> data;

	public Deque() {
		data = new LinkedList<E>();
	}

	public boolean isEmpty() {
		return data.size() == 0;
	}

	public void addFront(E val) {
		data.add(0, val);
	}

	public void addRear(E val) {
		data.add(val);
	}

	public E removeFront() {
		return data.remove(0);
	}

	public E removeRear() {
		return data.remove(data.size() - 1);
	}

	public E getFront() {
		return data.get(0);
	}

	public E getRear() {
		return data.get(data.size() - 1);
	}

	public static void main(String[] args) {
		Deque<Integer> deque = new Deque<>();
		deque.addFront(11);			// 11
		deque.addFront(22);			// 22 11
		deque.removeRear();			// 22
		deque.addRear(33);			// 22 33
		deque.addRear(44);			// 22 33 44
		deque.removeFront();		// 33 44
		System.out.println(deque.getFront());
		System.out.println(deque.getRear());
	}
}

File: code/Python/第5章 队列/5.5 双端队列/deque.py
class Deque:
	def __init__(self):
		self.data = []

	def is_empty(self):
		return not self.data

	def add_front(self, val):
		self.data.insert(0, val)

	def add_rear(self, val):
		self.data.append(val)

	def remove_front(self):
		if not self.is_empty():
			return self.data.pop(0)

	def remove_rear(self):
		if not self.is_empty():
			return self.data.pop()

	def get_front(self):
		if not self.is_empty():
			return self.data[0]

	def get_rear(self):
		if not self.is_empty():
			return self.data[len(self.data)-1]

def main():
	deque = Deque()
	deque.add_front(11)			# 11
	deque.add_front(22)			# 22 11
	deque.remove_rear()			# 22
	deque.add_rear(33)			# 22 33
	deque.add_rear(44)			# 22 33 44
	deque.remove_front()		# 33 44
	print(deque.get_front())
	print(deque.get_rear())

if __name__ == "__main__":
	main()
